Fix bugs on *nix (Linux in particular)

pgavlin opened the pull request:
With these changes, all of the Sockets tests except those that are (a) platform-specific, or (b) excluded due to differences in dual-stack semantics are passing on Linux.

pgavlin committed:
Fix some bugs on *nix.

- Always use nonblocking sockets + emulate blocking I/O if necessary.
  This is necessary to accommodate winsock's separation of blocking
  and overlapped I/O. TODO: respect send/receive timeouts.
- Fix repeated connect() behavior for Linux. Nonblocking sockets on
  Linux require that if a connect() attempt fails, the socket is
  disassociated from the previous address by a call to
  connect(AF_UNPSEC) before another attempt is made.

pgavlin committed:
Fix some bugs in S.N.Sockets{,.Tests}.

- A number of tests were reusing port numbers and so could not run in
  succession or in parallel
- Some tests were using the hostname "loopback", which is not mapped by
  default on *nix
- Sockets are expected to have dual stack behavior turned off by default
- AcceptAsync on *nix was not setting the remote address correctly.

pgavlin committed:
Timeout support.

pgavlin committed:
Fix various bugs with Unix sockets.

- poll() does not handle sockets that are close on other threads in a sane
  fashion, and so cannot be used to emulate blocking sockets. Use the
  existing async support instead. This entailed:
    - Overhauling AsyncOperation to support both events and callbacks as
      mechanisms for completion notification. Because event waits must be
      allowed to time out, this also involed adding cancellation support
      to AsyncOperation.
- SocketExceptionFactory had exposed a method that used the dangerous
  parameterless SocketException contructor. This method was removed.
- ConnectOverlappedAsyncResult was not setting the socket to connected

pgavlin committed:
Fix control message handling on Linux.

pgavlin committed:
Make the event loop thread a bit more robust.

pgavlin committed:
Fix synchronous I/O on *nix when the socket is non-blocking.

pgavlin committed:
Fix EPOLLHUP/SocketAsyncEvents.Close handling.

pgavlin committed:
Fix some issues with the Sockets tests.

- Mark Disconnect() and SendPackets() tests as Windows-specific
- Mark a few tests that cover dual-stack semantics with a dummy issue number
  while the differences are investigated
- Fix port numbers on a multitude of tests to avoid conflicts at runtime

pgavlin committed:
Undo some changes to run-test.sh.

pgavlin commented:
@CIPop PTAL at the Windows + test changes; @stephentoub PTAL at the cross-plat changes.

pgavlin commented:
I should point out that the performance tests don't yet run, as they contain Windows-specific code. I'm working on fixing this at the moment.

pgavlin committed:
Fix some issues exposed by the Sockets perf tests.

- There were two interesting races w.r.t. SocketAsyncEvents.{Close,HandleEvents}:
	- A Close() may complete after a call to epoll_wait/kevent64 returns with
      events for the closed context but before those events are handled. In this
      case, the GCHandle associated with the event will have been freed and
      GCHandle.Target will be null. Add a null check to fix this.
    - An event that causes queue stoppage may race with a call to Close(). If the
      event wins the race, Close() will attempt to operate on stopped queues.
      Add a check for stopped queues to fix this.

- There was a Windows-specific call to memcmp in the Sockets perf tests; this has been
  removed and replaced with a managed equivalent.

stephentoub commented on src/Common/src/System/Net/SafeCloseSocket.Unix.cs:
```diff
@@ -16,6 +16,9 @@ internal sealed partial class SafeCloseSocket :
 #endif
     {
+        private int _receiveTimeout = -1;
+        private int _sendTimeout = -1;
```
Do we want to use Timeout.Infinite?  I don't have a strong preference.

pgavlin replied:
I think I'll leave this as-is for now.

stephentoub commented on src/Common/src/System/Net/SafeCloseSocket.Unix.cs:
```diff
@@ -189,24 +211,37 @@ public static InnerSafeCloseSocket CreateSocket(AddressFamily addressFamily, Soc
+                        Interop.Sys.Close(fd);
+                        fd = -1;
+                    }
+                }
```
I'm probably just missing it, but where are we setting IsNonBlocking on the SafeHandle?

pgavlin replied:
`IsNonBlocking` is set by the `SocketPal.SetBlocking`. It does not reflect whether or not the underlying file descriptor is non-blocking. On *nix, we need to emulate Windows' separation of {non,}blocking I/O and overlapped I/O, which we do by setting every underlying socket to non-blocking and emulating blocking I/O as necessary based on the value of `SafeCloseSocket.IsNonBlocking`.

stephentoub commented on src/Common/tests/System.Net/Sockets/Performance/SocketTestClient.cs:
```diff
@@ -178,8 +178,8 @@ private void OnReceiveMessage()
             // Expect echo server.
-            if (memcmp(_sendBuffer, _recvBuffer, _sendBuffer.Length) != 0)
-            {
+			if (!Compare(_sendBuffer, _recvBuffer))
```
Nit: I think these are tabs rather than spaces

pgavlin replied:
Argh, I thought I had caught all of those. Thanks.

stephentoub commented on src/System.Net.Sockets.Legacy/tests/FunctionalTests/DisconnectAsync.cs:
```diff
@@ -15,6 +15,7 @@ public void OnCompleted(object sender, SocketAsyncEventArgs args)
         }
 
         [Fact]
+        [PlatformSpecific(PlatformID.Windows)]
```
Could you add a short comment next to the PlatformSpecific attribute indicating why it's platform-specific?  Is it because Disconnect isn't supported?

CIPop replied:
Better yet, instead of comments why not add a *Nix specific test with the expectation. (Throws NotSupportedException, etc)

pgavlin replied:
Both good suggestions.

@stephentoub these tests are Windows-specific b/c `Disconnect` cannot be supported on other platforms.

@CIPop I'll add the suggested tests.

stephentoub commented on src/System.Net.Sockets.Legacy/tests/FunctionalTests/DnsEndPointTest.cs:
```diff
@@ -58,8 +60,8 @@ public void Socket_ReceiveFromDnsEndPoint_ArgumentException()
             {
-                sock.Bind(new IPEndPoint(IPAddress.Loopback, 8080));
-                EndPoint endpoint = new DnsEndPoint("localhost", 8080);
+                sock.Bind(new IPEndPoint(IPAddress.Loopback, TestPortBase + 4));
```
Are these `+ 3`, `+ 4`, etc. being used in order to have different ports used for each test?  If so, that seems fragile and easy to get wrong, resulting in sporadic test failures.  Could we instead have a field or something that tracked the next available port, and do an Interlocked.Increment on it to get the ID for the test, or something like that?

CIPop replied:
Previously, all tests were using the same port. The problem appeared in the initial ToF attempt where tests were executed in parallel. 

The increment (interlocked or not) is still as fragile as this one as any process can take over the same port.

There is another method where you specify port 0 and the OS automatically binds to the next available ephemeral port. The problem is that the code will need to make a syscall to find the actual port (see the Security tests). This isn't really a common scenario except for FTP but solves this particular problem.

I think that most tests would work fine with this dynamic port allocation system although I would still leave a few on fixed ports.

pgavlin replied:
> There is another method where you specify port 0 and the OS automatically binds to the next available ephemeral port. The problem is that the code will need to make a syscall to find the actual port (see the Security tests). 

I thought about changing the tests to use this style, but decided against it for now. If nobody objects, I'll leave things as-is for now and open an issue to track the conversion.

CIPop replied:
No objections. Thank you for fixing the remaining tests to use different ports!

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -20,17 +20,104 @@ internal sealed class SocketAsyncContext
+                    switch ((State)state)
+                    {
+                        case State.Running:
+                            // A completion attempt is in progress. Keep busy-waiting.
```
This is a very tight spin loop.  Is this guaranteed to complete really quickly?  Consider using SpinWait.

pgavlin replied:
Good suggestion. I'll make the switch.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -39,7 +39,17 @@ public static SocketAsyncEngine Instance
+                                {
+                                    try
+                                    {
+                                        ((SocketAsyncEngine)o)._backend.EventLoop();
```
Consider moving the cast to be before the for loop so as to avoid needing to do it repeatedly (I realize we should only loop around again at this point if there was an exception, but still :smile:)

pgavlin replied:
Done.

pgavlin committed:
Fix some bugs on OS X.

- Define IPV6_PKTINFO on OS X
- Fix handling of EVFILT_EOF on write events
- Split lazy registration of read and write events

pgavlin committed:
Fix a typo.

CIPop commented on src/Common/tests/System.Net/Sockets/Performance/SocketTestClient.cs:
```diff
@@ -238,9 +238,23 @@ public Task<long> RunTest()
 
         protected abstract string ImplementationName();
-        
-        //TODO: Either find a different fast way to compare or extract this in Interop.
```
Please implement Compare as memcmp for Windows if this isn't possible for Linux.

In my investigations for WebSocket I noticed that this can have huge impact over the results and so we need to keep comparisons as fast as possible.

stephentoub replied:
Why is that important in the tests?  This meaningfully affects test running time?  Are we including this in whatever perf timing we're doing?

CIPop replied:
Yes, I expect this will meaningfully impact the test running time. 
We could think of ways to decouple correctness from performance measurements and remove this completely. Until then let's try to keep from diverging from the other variants (NetNative, Desktop).

Wouldn't a simple `#ifdef` block be enough to keep the old memcmp implementation for Windows only?

CIPop commented on src/System.Net.Sockets.Legacy/tests/FunctionalTests/DualModeSocketTest.cs:
```diff
@@ -8,6 +8,9 @@ namespace System.Net.Sockets.Tests
 {
     public class DualMode
     {
+        private const int DummyDualModeV6Issue = 8000;
```
Nit: in the 8000 ranges I originally thought this is a port. Please add a TODO explaining that these are issues that need to be opened.

Also: if there's any sort of tooling scanning for the ActiveIssue string outside of the compiler, it will probably fail not understanding the Issue ID.

CIPop commented on src/System.Net.Sockets.Legacy/tests/FunctionalTests/DualModeSocketTest.cs:
```diff
@@ -769,30 +772,30 @@ private void DualModeConnectAsync_IPEndPointToHost_Helper(IPAddress connectTo, I
                 ManualResetEvent waitHandle = new ManualResetEvent(false);
                 SocketAsyncEventArgs args = new SocketAsyncEventArgs();
                 args.Completed += new EventHandler<SocketAsyncEventArgs>(AsyncCompleted);
-                args.RemoteEndPoint = new DnsEndPoint("loopback", TestPortBase + 51);
```
Just curious: Linux doesn't translate `loopback` to the loop-back IP? Windows does according to `ping loopback`.

pgavlin replied:
Correct: most non-Windows systems have no entry in `/etc/hosts` (or whatever else is used for static name resolution) for `loopback` by default.

CIPop commented on src/System.Net.Sockets.Legacy/tests/FunctionalTests/SendPacketsAsync.cs:
```diff
@@ -51,6 +51,7 @@ public SendPacketsAsync(ITestOutputHelper output)
         #region Basic Arguments
 
         [Fact]
+        [PlatformSpecific(PlatformID.Windows)]
```
I noticed that none of the tests in this file apply to non-Windows platforms. It would be interesting to understand why: maybe a TODO item to add Unix tests that clarify the behavior?

CIPop replied:
If none of the tests apply for Nix, shouldn't we only include it for Windows in the csproj?

pgavlin replied:
> It would be interesting to understand why: maybe a TODO item to add Unix tests that clarify the behavior?

None of these apply to non-Windows platforms because the API underlying `SendPackets` is peculiar to Winsock. These should probably instead be excluded on non-Windows platforms with an issue to track whether or not we need to provide support for these APIs on those platforms.

> If none of the tests apply for Nix, shouldn't we only include it for Windows in the csproj?

This is debatable. In this case, no, due to the aforementioned uncertainty re: the necessity of the APIs, but perhaps in the case of Disconnect (where there is no possible implementation on other platforms) we should exclude those files.

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -114,6 +114,11 @@ public Socket(AddressFamily addressFamily, SocketType socketType, ProtocolType p
 
+            if (addressFamily == AddressFamily.InterNetworkV6)
+            {
+                DualMode = false;
```
This shouldn't be necessary. If for some reason it is for *NIX, we should move it within the PAL. 
As written, the above ctor will call `setsockopt` twice with conflicting values: was that the intent?

pgavlin replied:
> This shouldn't be necessary. If for some reason it is for *NIX, we should move it within the PAL. 

The initial value of IPV6_V6ONLY may differ between platforms. I thought about moving it behind the PAL and ended up leaving it here for clarity. I'll change this to a `Debug.Assert()` and move the setup behind the PAL.

CIPop replied:
I'm concerned that changing to `Debug.Assert()` might impact execution in debug mode since it makes a call to a Win32 API. Is there any kind of in-memory state that we can probe without calling an API?

pgavlin committed:
Address "multiple connect" calls and SocketAsyncContext robustness.

- Support "multiple connect" operations (e.g. Connect(IPAddress[]), DnsEndPoint connections,
  etc.) on platforms that do not support multiple calls to connect() on the same socket.
  For these platforms, each address will be tried on a new socket. Once a good address is
  found (if any), the user's socket will attempt to connect to that address.
- Improve SocketAsyncContext's robustness in the case of contexts which have only registered
  for Read events.

pgavlin committed:
Respond to non-test CR feedback.

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:
```diff
@@ -530,350 +581,792 @@ public static unsafe SafeCloseSocket CreateSocket(SocketInformation socketInform
+// Disable CS0162: Unreachable code detected
+//
+// SuportsMultipleConnectAttempts is a constant; when false, the following lines will trigger CS0162.
+#pragma warning disable 162
```
Instead of a code constant can't we have a `#define` and compile this section out?
It would remove the need for the `Debug.Fail()` within `PrimeForNextConnectAttempt` in OSX.

pgavlin replied:
We could, yes. I don't have terribly strong feelings here (despite a deep loathing for `#ifdefs`), but it was my understanding that we're trying to avoid `#define` soup in CoreFX. @stephentoub, any guidance?

stephentoub replied:
I'd prefer either the pragma or a partial method to a #if.

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -1053,16 +1055,71 @@ public void Connect(IPAddress[] addresses, int port)
+#pragma warning disable 162
             Exception lastex = null;
-            foreach (IPAddress address in addresses)
+            if (SocketPal.SupportsMultipleConnectAttempts)
```
Having the compiler warning is a code smell: why not have this extracted within PAL?
Is there any use for any of the Windows implementations to have the `else` side of this check?

pgavlin replied:
> why not have this extracted within PAL?

The algorithm is not exactly platform-dependent, it's capability-dependent. All a new platform has to do for this code is declare is whether or not it supports multiple connection attempts on a socket--moving it behind the PAL would likely require a fair amount of code copying. I'd be more inclined to guard this code with `#ifdef`s than to move it behind the PAL.

CIPop replied:
> All a new platform has to do for this code is declare is whether or not it supports multiple connection attempts on a socket--moving it behind the PAL would likely require a fair amount of code copying. I'd be more inclined to guard this code with #ifdefs than to move it behind the PAL.

I don't agree with the code copying part: at this time only one platform has this issue. Future extensions can be easily combined at an intermediate PAL layer: Windows (with 2 platforms), Linux will share the current code. OSX (and probably FreeBSD) will share the other classes.

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -1053,16 +1055,71 @@ public void Connect(IPAddress[] addresses, int port)
+            }
+            else
+            {
+                EndPoint endpoint = null;
```
Regarding above comment: if my understanding is correct, the following code will never execute on a Windows device. (IoT, Phone, etc will still compile and distribute it, making the footprint larger.)

pgavlin replied:
> Regarding above comment: if my understanding is correct, the following code will never execute on a Windows device.

That's correct.

> (IoT, Phone, etc will still compile and distribute it, making the footprint larger.)

Ideally only as IL: any reasonable compiler should eliminate this from the native code.

CIPop replied:
Actually, I was talking about disk footprint. Decreasing it was the primary driver for CoreFX, at least in the initial discussions.

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -2208,35 +2265,33 @@ public object GetSocketOption(SocketOptionLevel optionLevel, SocketOptionName op
+            {
                 //
-                // if the native call fails we'll throw a SocketException
+                // update our internal state after this socket error and throw
```
Nit: capitalize comment and end in "." (Although I presume one of us will have to go through the rest of the code and do that for cleanup.)

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -4755,7 +4810,9 @@ public static bool ConnectAsync(SocketType socketType, ProtocolType protocolType
                 {
-                    multipleConnectAsync = new MultipleSocketMultipleConnectAsync(socketType, protocolType);
+                    multipleConnectAsync = SocketPal.SupportsMultipleConnectAttempts ?
+                        (MultipleConnectAsync)(new DualSocketMultipleConnectAsync(socketType, protocolType)) :
```
I see that the the following has been renamed but I don't understand the reason:
`internal class MultipleSocketMultipleConnectAsync : MultipleConnectAsync`
`+    internal sealed class DualSocketMultipleConnectAsync : MultipleConnectAsync`

CIPop replied:
Ah, actually `DualSocketMultipleConnectAsync` is a new class but GitHub diff is making a mess. 
I still don't understand why all these SocketPal.SupportsMultipleConnectAttempts exist: the Windows and even Linux code supports it. Why are we getting so much code churn to support OSX instead of pushing this code in PAL?

I think at this point we'd need some sort of architecture design that explains these changes. The CR became too complicated to understand the intent.

pgavlin replied:
> Why are we getting so much code churn to support OSX instead of pushing this code in PAL?

It's not just churn to support OS X specifically: it's churn to enable _any_ platform that does not support multiple connect attempts on the same socket (e.g. some/all of the BSDs).

> I think at this point we'd need some sort of architecture design that explains these changes. The CR became too complicated to understand the intent.

I'll split these changes out into their own PR.

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -6083,7 +6141,7 @@ private static bool DoDnsCallback(IAsyncResult result, MultipleAddressConnectAsy
         }
 
-        private class MultipleAddressConnectAsyncResult : ContextAwareResult
+        private sealed class MultipleAddressConnectAsyncResult : ContextAwareResult
```
What is the reason to seal the class?

During work on NetworkInformation I noticed the same changes that had to be reverted.

https://msdn.microsoft.com/en-us/library/ms229023(v=vs.110).aspx
**X DO NOT** seal classes without having a good reason to do so.

pgavlin replied:
> What is the reason to seal the class?

It's a private type that AFAICT is not an extension point.

> During work on NetworkInformation I noticed the same changes that had to be reverted.

What caused the reversion? Was there code outside of NetworkInformation that was inheriting from the type?

> https://msdn.microsoft.com/en-us/library/ms229023(v=vs.110).aspx
> X DO NOT seal classes without having a good reason to do so.

I believe that this guidance is out of date. @stephentoub ?

stephentoub replied:
Whether or not it's out of date, it's meant to be about publically-exposed types, which this is not.  For any internal-only types, in general I believe we should have a seal-by-default stance, and only unseal when needed.  Sealing has potential performance benefits as well as potential usability benefits around understanding the extensibility potential of a type.

CIPop replied:
> What caused the reversion? Was there code outside of NetworkInformation that was inheriting from the type?

`SafeCloseSocket` is extended by `SafeCloseSocketAndEvent`. There were a few other changes that will bubble up to System.Net.Primitives. These were expected and luckily not that complicated given that we've started working on PAL before we completely understand how these types are layered.

> It's a private type that AFAICT is not an extension point.
> For any internal-only types, in general I believe we should have a seal-by-default stance
> Sealing has potential performance benefits as well as potential usability benefits around understanding the extensibility potential of a type.

Makes sense. Let's add it to the coding style rules and make the code uniform (if we change this type, let's change others in a similar way).

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/_MultipleConnectAsync.cs:
```diff
@@ -31,6 +39,10 @@ private enum State
 
         private object _lockObject = new object();
 
+        protected abstract bool RequiresUserConnectAttempt { get; }
```
Instead of changing the global code to accept various OS behaviors, we should instead push the behavior down within PAL and conditionally compile based on what the platform supports.

pgavlin replied:
Are you suggesting that this entire class (the entire file, really) be duplicated per-platform? The code needs to accommodate an extra state in its state machine depending on whether or not the underlying platform supports multiple connection attempts on a single socket. The only way that I can think of to avoid the extra state is to try _every_ connect on a new socket first on platforms that don't support multiple connection attempts, which adds significant overhead (for questionable benefit).

stephentoub replied:
I've been trying to follow along with the review; I see this discussion of the PAL going on in a few different places, so I'll respond here, but it applies to the others as well.

In general, across corefx, I think we've done a good job of keeping the separation between common code and platform-specific code clean.  We've done so using a variety of patterns, as there isn't one that fits all cases.

In general, if we have a static class that needs some platform-specific functionality, we've either created a separate XxPal class that the common class calls into, or we've created partial classes to contain the static methods.  We initially started down the path of the former, but it's difficult to do well when you get to instance members and state, which pushes you down the path of the partial class design.  So the majority of corefx is on the partial class plan.

However, there have been some cases where neither of those approaches is perfectly feasible, and the best solution is to have some PAL-based capability exposed that the common code can then make decisions based on.  For example, with different file systems handling casing differently, having to duplicate functionality in the different PAL implementations based on whether the file system is case sensitive or not leads to tons of duplication.  Instead, it's much cleaner to expose a constant / readonly static from the PAL layer and let the common code branch based on that.  In most cases, the unfollowed branch doesn't even end up in the IL, as Roslyn will do dead branch elimination based on constants.  And for readonly statics like Booleans, the JIT will do dead branch elimination in the generated assembly. Another example where we hit this was in memory mapped files.  When using CreateNew, we need to create an anonymous backing store for the created map.  We'd really like to do so using an in-memory object (shm_open), but even though that function exists on both Linux and OS X, it doesn't have the charactersitics we need on OS X.  As such, we have a backup solution that is to create a file-based object.  For a variety of reasons, we're now structured so that at run time the PAL tells the common code which approach to use, and the common code branches based on that.

This is all to say that there are definitely cases where it's much cleaner IMHO to have the common code do a bit of branching based on checking capabilities exposed from the PAL.  In fact, I wish we took that path in a few more cases and avoided some code duplication that's led to some bugs.  For example, when I originally did the Unix implementation of FileStream, there were these kinds of branches, and so I ended up essentially just copying the Win32 implementation to a new file, deleting stuff that wasn't relevant, and adding in the new Unix-specific stuff.  In retrospect that was a mistake, and at some point I'd like to go revisit it.  Overtime, we've found a bug here and a bug there where because of the code copying, the implementations had drifted apart and were behaving more differently than they should have been.  When we've had to make a bug fix, we've had to do it in two different places.  Etc.

I agree it would be nice if these capability checks weren't necessary.  From a brief look over the code, though, it seems like it could be the least bad option.

My $.02.

CIPop commented on src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:
```diff
@@ -111,6 +111,7 @@ private void DualModeConnectAsync_IPEndPointToHost_Helper(IPAddress connectTo, I
                 {
                     throw new SocketException((int)args.SocketError);
                 }
+                Assert.True(server.WaitHandle.WaitOne(5000), "Timed out while waiting for accept");
```
Why is this assert required?

CIPop replied:
NIT: new line missing after closing brace.

pgavlin replied:
I think I added this while debugging the tests; I'll remove it.

CIPop commented on src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:
```diff
@@ -122,35 +123,36 @@ private void DualModeConnectAsync_IPEndPointToHost_Helper(IPAddress connectTo, I
                 socket.ConnectAsync(args);
 
                 waitHandle.WaitOne();
+                Assert.True(server.WaitHandle.WaitOne(5000), "Timed out while waiting for accept");
```
I notice that there are a lot of these asserts to wait for the server-side accept. They are changing the semantics of the tests.

pgavlin replied:
See above--I'll remove this.

CIPop commented on src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:
```diff
@@ -77,31 +77,31 @@ public void ConnectAsyncV4IPEndPointToV6Host_Fails()
         {
-            Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
-            using (SocketServer server = new SocketServer(_log, listenOn, dualModeServer, TestPortBase + 13))
+            using (Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp))
```
If we want to change these tests to use the proper `using` statements, we're going to need special finalization tests written, where socket handles are closed by the GC at a later time instead of getting forcefully disposed.

pgavlin replied:
I'll revert this change.

CIPop commented on src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:
```diff
@@ -350,11 +353,11 @@ public void Socket_SendToAsyncV4IPEndPointToV4Host_Throws()
             bool async = socket.SendToAsync(args);
             Assert.False(async);
-            Assert.Equal(SocketError.Fault, args.SocketError);
+			Assert.NotEqual(SocketError.Success, args.SocketError);
```
This is a clear behavior change. We'll need to document it properly as certain applications will take dependencies on exact return code. 
Please add a comment or change the test to outline what each OS is expecting to return.

pgavlin replied:
I'll revert this to its original form and mark this with an issue. We'll need to do the work to return the same error here if possible.

CIPop commented on src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:
```diff
@@ -520,31 +523,31 @@ public void ReceiveFromAsyncV4BoundToSpecificV6_NotReceived()
+                SocketUdpClient client = new SocketUdpClient(_log, serverSocket, connectTo, port);
                 if (async && !waitHandle.WaitOne(200))
                 {
                     throw new TimeoutException();
```
As written the `TestFinalizers()` test isn't going to have any effect anymore.

CIPop commented on src/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs:
```diff
@@ -633,6 +638,10 @@ public void Dispose()
                     server.Dispose();
+                    if (accepted != null)
+                    {
+                        accepted.Dispose();
```
See above: while this is the right pattern, it won't test for how a lot of code is actually written.

If we fix this test, please add tests that will verify that disposing the accept socket handles are handled well by the GC finalizer (not only when properly Disposed by the application code).

pgavlin replied:
I'll revert this.

pgavlin commented:
I'm going to close this PR and split it into multiple PRs: one for the tests, one or two for some of the more significant *nix changes, and one for the multiple-connect change. The lattermost may have to wait until the PRs for the earlier *nix changes are merged, as it builds upon them.
